src/lib.rs

```diff
@@ -643,8 +650,17 @@ mod implicit {
         let mut last_strong_is_l = false;
         let mut et_run_indices = Vec::new(); // for W5
 
-        let mut indices = sequence.runs.iter().flat_map(Clone::clone).peekable();
+        // Like sequence.runs.iter().flat_map(Clone::clone), but make indices itself clonable.
+        fn id(x: LevelRun) -> LevelRun { x };
+        let mut indices = sequence.runs.iter().cloned().flat_map(id as fn(LevelRun) -> LevelRun);
+
         while let Some(i) = indices.next() {
+            // "Apply rules W1–W7, N0–N2, and I1–I2, in the order in which they appear below,
+            //  to each of the isolating run sequences, applying one rule to all the characters
+            //  in the sequence in the order in which they occur in the sequence
+            //  before applying another rule to any part of the sequence."
+            // At the moment we do all of W1-W7 in one pass. Is this OK?
+            // FIXME: If so, add a comment explaining why.
```

Comment:
I think you're right.  Looking at the [Gecko implementation][1], it optimizes these into a single pass but preserves extra state (e.g. in `struct Isolate`) to make this work.  Perhaps we should break this out into separate loops for each W*n* rule for now (as in the spec), and optimize it more carefully later.  Unfortunately, this will greatly increase the number of times we loop over the entire input. :(

[1]: https://hg.mozilla.org/mozilla-central/file/e7e69cc8c07b/layout/base/nsBidi.cpp#l1119